packages/frontend/core/src/modules/open-in-app/services/index.ts

```diff
@@ -0,0 +1,103 @@
+import { WorkspaceFlavour } from '@affine/env/workspace';
+import type { GlobalState, WorkspacesService } from '@toeverything/infra';
+import { LiveData, OnEvent, Service } from '@toeverything/infra';
+
+import { resolveLinkToDoc } from '../../navigation';
+import { WorkbenchLocationChanged } from '../../workbench/services/workbench';
+
+const storageKey = 'open-link-mode';
+
+export enum OpenLinkMode {
+  ALWAYS_ASK = 'always-ask', // default
+  OPEN_IN_WEB = 'open-in-web',
+  OPEN_IN_DESKTOP_APP = 'open-in-desktop-app',
+}
+
+@OnEvent(WorkbenchLocationChanged, e => e.onNavigation)
+export class OpenInAppService extends Service {
+  private initialized = false;
+
+  private initialUrl: string | undefined;
+
+  readonly showOpenInAppBanner$ = new LiveData<boolean>(false);
+  readonly showOpenInAppPage$ = new LiveData<boolean | undefined>(undefined);
+
+  constructor(
+    public readonly globalState: GlobalState,
+    public readonly workspacesService: WorkspacesService
+  ) {
+    super();
+  }
+
+  onNavigation() {
+    // check doc id instead?
+    if (window.location.href === this.initialUrl) {
+      return;
+    }
+    this.showOpenInAppBanner$.next(false);
+  }
+
+  /**
+   * Given the initial URL, check if we need to redirect to the desktop app.
+   */
+  bootstrap() {
+    if (this.initialized || !window) {
+      return;
+    }
+
+    this.initialized = true;
+    this.initialUrl = window.location.href;
+
+    const maybeDocLink = resolveLinkToDoc(this.initialUrl);
+    let shouldOpenInApp = false;
+
+    if (
+      maybeDocLink &&
+      this.workspacesService.list.workspaces$.value.find(
+        w => w.id === maybeDocLink.workspaceId
+      )?.flavour === WorkspaceFlavour.AFFINE_CLOUD
```

Comment:
Maybe the user is not logged in at this time?
We just need to check if the workspace is not a local workspace.

Reply:
I see, this list only makes sense if the user is logged in? I will read the value from localstorage instead.